### PR TITLE
Add tests for S2I Perl and Python images with Hot Deploy

### PR DESCRIPTION
**Pull request by mnagy**
@jhadvig

### CODE REVIEW

**Comment by mnagy** [comment]
I've decided to add in Python as well since it depends on the helper functions.

[test][extended:core]

File: test/extended/images/helper.go [hunk 1]
```diff
@@ -16,3 +24,52 @@ func GetImageLabels(c client.ImageStreamImageInterface, imageRepoName, imageRef
...
+
+	pod, err := oc.KubeREST().Pods(oc.Namespace()).Get(pods[0])
+	if err != nil {
+		return err
+	}
+	return oc.Run("exec").Args(pod.Name, "-c", pod.Spec.Containers[0].Name, "--", "sed", "-ie", sedScript, file).Execute()
+}
+
+// GetPageCount makes a http request for an example application and returns the page count
+func GetPageCount(oc *exutil.CLI, endpoint string, re *regexp.Regexp) (int, error) {
```
Comment (jhadvig) [inline test/extended/images/helper.go]: Would it be better to swap the regexp with a string ? Dont se any benefit of having a regular here, since we would like to reuse this method also for other framework

File: test/extended/images/helper.go [hunk 2]
```diff
@@ -16,3 +24,52 @@ func GetImageLabels(c client.ImageStreamImageInterface, imageRepoName, imageRef
 	}
 	return image.Image.DockerImageMetadata.Config.Labels, nil
 }
+
+// ModifySourceCode will modify source code in the pod of the application
+// according to the sed script.
```
Comment (jhadvig) [inline test/extended/images/helper.go]: s/sed script/sed expression/

File: test/extended/images/helper.go [hunk 3]
```diff
@@ -16,3 +24,52 @@ func GetImageLabels(c client.ImageStreamImageInterface, imageRepoName, imageRef
...
+	}
+	if len(pods) != 1 {
+		return fmt.Errorf("Got %d pods for selector %v, expected 1", len(pods), selector)
+	}
+
+	pod, err := oc.KubeREST().Pods(oc.Namespace()).Get(pods[0])
+	if err != nil {
+		return err
+	}
+	return oc.Run("exec").Args(pod.Name, "-c", pod.Spec.Containers[0].Name, "--", "sed", "-ie", sedScript, file).Execute()
```
Comment (mfojtik) [inline test/extended/images/helper.go]: don't need to use container name, if you specify a pod it should pick the first container by default AFAIK

File: test/extended/images/helper.go (hunk 1 above)
Comment (mnagy) [inline test/extended/images/helper.go]: Yes, will do.

File: test/extended/images/helper.go (hunk 3 above)
Comment (mnagy) [inline test/extended/images/helper.go]: I know, but you will get a warning, which is annoying.

File: test/extended/images/helper.go (hunk 2 above)
Comment (mnagy) [inline test/extended/images/helper.go]: I think it doesn't matter, either one is fine. From sed(1):
```
       -e script, --expression=script

              add the script to the commands to be executed
```

**Comment by mnagy** [comment]
Addressed comments, fixed Python tests and rebased for good measure.

PTAL

**Comment by mnagy** [comment]
Python and Perl hot deploy test work, search Jenkins logs for "hot deploy".

File: test/extended/images/helper.go [hunk 4]
```diff
@@ -1,12 +1,20 @@
...
+	"net/http"
+	"time"
+
 	"github.com/openshift/origin/pkg/client"
 	"github.com/openshift/origin/pkg/image/registry/imagestreamimage"
+	exutil "github.com/openshift/origin/test/extended/util"
+	"k8s.io/kubernetes/pkg/labels"
 )
 
-// GetImageLabels retrieves Docker labels from image from image repository name and
```
Comment (jhadvig) [inline test/extended/images/helper.go]: Is this change necessary ?

Comment (mnagy) [inline test/extended/images/helper.go]: No. No idea how that change got there, fixed.

File: test/extended/images/helper.go [hunk 5]
```diff
@@ -16,3 +24,48 @@ func GetImageLabels(c client.ImageStreamImageInterface, imageRepoName, imageRef
...
+// CheckPageContains makes a http request for an example application and checks
+// that the result contains given string
+func CheckPageContains(oc *exutil.CLI, endpoint, contents string) (bool, error) {
+	address, err := exutil.GetEndpointAddress(oc, endpoint)
+	if err != nil {
+		return false, err
+	}
+
+	// Make the request
+	resp, err := http.Get(fmt.Sprintf("http://%s/", address))
```
Comment (mfojtik) [inline test/extended/images/helper.go]: I think @bparees has some function to query the HTTP endpoint already for example tests.

Comment (mfojtik) [inline test/extended/images/helper.go]: also I think this need some retry logic

Comment (mnagy) [inline test/extended/images/helper.go]: Why retry logic? The pod should be ready at this point. Problem with retrying would be that the count could be different than expected.

Comment (bparees) [inline test/extended/images/helper.go]: https://github.com/openshift/origin/blob/master/test/extended/util/framework.go#L446

Comment (mfojtik) [inline test/extended/images/helper.go]: Does the Pod have readiness check that makes sure the pod is "ready" ?

Comment (bparees) [inline test/extended/images/helper.go]: https://github.com/openshift/origin/blob/master/test/extended/images/sample_repos.go#L63-L74

Comment (mnagy) [inline test/extended/images/helper.go]: @mfojtik yes. CheckPageContains() calls GetEndpointAddress() which calls WaitForAnEndpoint(). That is equivalent to the code @bparees linked above.

**Comment by mnagy** [comment]
@mfojtik addressed your comments and rebased.

File: test/extended/images/helper.go (hunk 5 above)
Comment (bparees) [inline test/extended/images/helper.go]: @mnagy you should definitely be wrappering the existing function here.  if you want to add response code checking to the existing function, that's fine.

Comment (mnagy) [inline test/extended/images/helper.go]: Ack, will do.

File: test/extended/images/helper.go [hunk 6]
```diff
@@ -16,3 +24,48 @@ func GetImageLabels(c client.ImageStreamImageInterface, imageRepoName, imageRef
 	}
 	return image.Image.DockerImageMetadata.Config.Labels, nil
 }
+
+// ModifySourceCode will modify source code in the pod of the application
+// according to the sed script.
+func ModifySourceCode(oc *exutil.CLI, selector labels.Selector, sedScript, file string) error {
```
Comment (bparees) [inline test/extended/images/helper.go]: this feels oddly scoped to me.  why not have this method take a pod or a container, instead of a label, given that it only works if the thing it finds is a single pod and it only actualy operates on the first container in that pod?

Comment (mnagy) [inline test/extended/images/helper.go]: good point

Comment (mnagy) [inline test/extended/images/helper.go]: I've looked at it again and it uses the WaitForPods() function which takes a label, the point being that the function will also make sure the pod is running. I could move the waiting-until-pod-is-running logic outside of that function, but I think it will only make the tests look less comprehensible.

**Comment by mnagy** [comment]
FYI, Travis failing for unrelated reasons. I checked Jenkins logs and hot deploy tests are succeeding.

File: test/extended/util/framework.go [hunk 7]
```diff
@@ -452,6 +452,9 @@ func FetchURL(url string, retryTimeout time.Duration) (response string, err erro
 			// because otherwise it's going to give up.
 			return false, nil
 		}
+		if r.StatusCode != 200 {
+			return false, fmt.Errorf("Got HTTP response %d, expected 200", r.StatusCode)
```
Comment (bparees) [inline test/extended/util/framework.go]: I think you need to return nil here.  if you return an error, the wait.Poller is going to stop polling.

in the case that an app server returns a 404 until the app is up, you actually do want to keep polling.

**Comment by mnagy** [comment]
Addressed @bparees's comment. Additionally, I've added a wait for build, because I discovered that WaitForAnEndpoint() will hang if the build fails. And one minor nit for @jhadvig.

**Comment by bparees** [comment]
lgtm.  what is the nit for @jhadvig ?

**Comment by mfojtik** [comment]
LGTM as well

**Comment by mnagy** [comment]
@bparees adding the postfix parameter here: https://github.com/openshift/origin/blob/7be12f6985226449a7d09f3733542fc5eba9bcac/test/extended/images/helper.go#L46

File: test/extended/images/helper.go (hunk 3 above)
Comment (jhadvig) [inline test/extended/images/helper.go]: Now when I think about it, wouldnt it be better to be able to run any kind of command here ? and not just sticking to sed ?

File: test/extended/images/helper.go [hunk 8]
```diff
@@ -16,3 +22,36 @@ func GetImageLabels(c client.ImageStreamImageInterface, imageRepoName, imageRef
...
+	pod, err := oc.KubeREST().Pods(oc.Namespace()).Get(pods[0])
+	if err != nil {
+		return err
+	}
+	return oc.Run("exec").Args(pod.Name, "-c", pod.Spec.Containers[0].Name, "--", "sed", "-ie", sedScript, file).Execute()
+}
+
+// CheckPageContains makes a http request for an example application and checks
+// that the result contains given string
+func CheckPageContains(oc *exutil.CLI, endpoint, postfix, contents string) (bool, error) {
```
Comment (bparees) [inline test/extended/images/helper.go]: "postfix"?  path?

**Comment by bparees** [comment]
@mnagy needs a squash

File: test/extended/images/helper.go (hunk 3 above)
Comment (bparees) [inline test/extended/images/helper.go]: @jhadvig let's cross that bridge when we have a use case.

Comment (jhadvig) [inline test/extended/images/helper.go]: @bparees I've already have one(deleting `public/index.html` in rails), but sure, I can update it in the following PR I have prepared.

Comment (mfojtik) [inline test/extended/images/helper.go]: if you have that use-case, wait for this to merge and then change the function to work the way you want.

**Comment by jhadvig** [comment]
LGTM ... lets merge it !

**Comment by bparees** [comment]
did the extended tests pass jenkins?

**Comment by jhadvig** [comment]
not sure, will re-launch the image ext-tests
[test][extended:core(images)]

**Comment by mnagy** [comment]
@bparees you can see in the jenkins console log that Hot Deploy tests passed (search for "hot deploy"). @jhadvig's tag didn't work, maybe he was too late. I've edited my previous comment that didn't include focus.

**Comment by bparees** [comment]
@mnagy i'm asked you because i don't want to dig through it since you know the TCs you expect to see run/succeed better than i do.

[merge]

**Comment by mnagy** [comment]
@bparees sure, no problem. I just wanted to also give you an easy way to verify I'm not lying, if you wanted to check :)

**Comment by bparees** [comment]
@mnagy i trust you..........for now :)

**Comment by jhadvig** [comment]
:+1:
